File: Data Structures & Algorithms/combinations/submission-3.cpp
class Solution {
private:
    void dfs(int i, vector<int>& curr, vector<vector<int>>& out, int n, int k) {
        if (curr.size() == k) {
            out.push_back(curr);
            return;
        }

        if (i > n) return;

        curr.push_back(i);
        dfs(i + 1, curr, out, n, k); // include

        curr.pop_back();
        dfs(i + 1, curr, out, n, k); // exclude
    }
    
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> out;
        vector<int> curr;
        dfs(1, curr, out, n, k);
        return out;
    }
};

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-3.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
            
        digits_to_letters = {
            "2" : "abc", "3" : "def", "4" : "ghi", 
            "5" : "jkl", "6" : "mno", "7" : "pqrs",
            "8" : "tuv", "9" : "wxyz"
        }

        letter_candidates = []
        for digit in digits:
            letter_candidates.append(digits_to_letters[digit])

        out = []
        curr = []

        def dfs(i):
            if i == len(digits):
                out.append("".join(curr))
                return
            
            ith_letters = letter_candidates[i]
            for letter in ith_letters:
                curr.append(letter)
                dfs(i + 1)
                curr.pop()
        
        dfs(0)
        return out